doc/remailers.rst

```diff
@@ -0,0 +1,125 @@
+and Autocrypt-enabled subscribers, but becomes increasingly ineffective with
+larger numbers of passive or non Autocrypt-enabled subscribers. As with any
+encrypted group communication, the benefits of end-to-end encryption also dilute
+with larger numbers of recipients.
+
+For compatibility with clients that are not Autocrypt-capable, the remailer
+software might want to offer manual configuration of keys, e.g. via a web
+interface.
+
+Injecting Reamiler Autocrypt-Headers
+~~~~~
+
+To inform subscribers of its key, the remailer injects an Autocrypt-header into
+outgoing messages. The header format is the same as usual, where the ``addr``
+attribute is the address of the list, and ``keydata`` is their Autocrypt key.
+
+In addition to those two attributes, the remailer SHOULD add the
+``_list-recommendation`` attribute, and MAY add ``_list-recipients`` attribute.
+Both of these attributes are exclusive for use by remailers, and MUST NOT be
+used by regular mailbox addresses. Both of these attributes are non-critical
+(starting with an underscore), which will cause an Autocrypt-capable client that
+does not handle these attributes to skip over them, falling back to the usual
+Autocrypt behavior.
+
+For example::
+
+    Autocrypt: addr=post@list.example; _list-recommendation=mandatory; keydata=XXX
+    Autocrypt: addr=post@list.example; _list-recommendation=available; _list-recipients=7/15; keydata=XXX
+
+The ``_list-recommendation`` attribute SHOULD be included in an Autocrypt header
+sent by a remailer. Its value may be either ``available`` or ``mandatory``. If
+the value is set to ``mandatory``, the remailer requires encryption of every
+message, and will reject plaintext messages. If the value is ``available``,
+encryption is offered as an opt-in mechanism and plaintext messages will be
+handled as usual. In either case, encrypted messages will be delivered
+exclusively to subscribers for whom an encryption key is available.
+
+To inform the decision of whether to encrypt or not, the remailer MAY include a
+``_list-recipients`` attribute. This attribute contains the current number of
+subscribers for which an encryption key is available, and the total number of
+subscribers, separated by a slash character ('/'). This attribute is only useful
+if the ``_list-recommendation`` attribute is set to ``available``.
+
+Administrative Considerations
+----
+
+Graceful Up- and Downgrade
+~~~
+
+To allow graceful up- and downgrade of lists with no intervention by the list
+administrator, a remailer software distribution could automatically include
+Autocrypt-headers with outgoing mail depending on the number of
+Autocrypt-enabled subscribers. For example, a remailer software distribution may
+choose to inject Autocrypt headers only if encryption keys are available for a
+majority of subscribers. Note that obviously, end-to-end encryption is
+unsuitable for lists that are publicly archived.
+
```

Comment:
Maybe add an additional caveat here about wanting some hysteresis or other kinds of smoothing -- one person's key becoming unavailable and then available shouldn't cause flip-flopping of the presence of the header on subsequent messages.

Reply:
It doesn't actually flip-flop, unless there is a 35 days delay between the messages (and even then it only becomes "discouraged") :)

Reply:
it could flip flop if a new user joins, or if a user quits and the number generated is based on a proportion of subscribers, right?  or a new user subscribes (without an autocrypt key) and then later adds their autocrypt key.  

My point is that we ought to explicitly think about all of the use cases and ensure that we are ok with this kind of dynamic threshhold-setting.

alternately, we could just emit the list's autocrypt header as long as the remailer thinks at least one recipient is autocrypt-capable and let the clients decide based on the `_list-recommendation`.

if we really want to ensure that peers only send to mailing lists if they can deal with this kind of decision-making, we could also convert `_list-recommendation` to `list-recommendation` (making it critical, so that normal autocrypt clients would ignore the header until they've been taught about encrypted re-mailers).

Reply:
I meant that while inclusion of the header will flip-flop, the calculated autocrypt-recommendation on clients will not due to the way timestamps are handled. If it didn't that would be annoying for the user even for the non-mailinglist case. Is this included in your consideration?